### PR TITLE
[MPS] Migrate var/std to native Metal kernels (Welford)

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
Part of #187455.

## Stack Position / Review Order

This is the third reduce-family split and should be reviewed after:

1. #188156: shared `value_reduction<Op, Load, TA>` base for the sum family.
2. #188333: `prod` on top of that shared base.
3. This PR: `var` / `std` / `var_mean` / `std_mean` native Metal Welford kernels.

This rebuild intentionally drops the stale shared `uint32 -> int64` reduction-indexing migration from the earlier #187787 draft. The native Welford path keeps uint32 indexing in the existing `NormParams` domain and rejects unsupported `>2^32` input/output/reduction cases with explicit errors instead of widening the shared reduction infrastructure in this PR.

## Summary

Replaces the MPSGraph-backed `var`, `std`, `var_mean`, and `std_mean` reductions on MPS with native Metal Welford kernels. The implementation keeps Welford-specific kernels separate from the shared sum/prod/value reduction path, preserving the reduce-family stack split while removing another per-shape MPSGraph graph-cache path.

The branch includes:

- generic Welford reduction for arbitrary dims using existing uint32 `NormParams` indexing;
- specialized all-reduce, outer-dim, inner-dim, wide-inner, and thin-inner Welford kernels for representative normalization shapes;
- complex input handling through real/imag component reductions, preserving real-valued variance/std plus complex mean semantics without adding complex Welford Metal kernels;
- OpInfo/common_mps updates for the newly supported MPS var/std dtype surface;
- focused MPS regression coverage for correction, duplicate dims, scalar dims, partial-simdgroup dispatch, large-offset numerical stability, non-stride-aligned wide tails, and GroupNorm-relevant shapes.

Coverage owners: OpInfo / `common_methods_invocations.py` and `torch/testing/_internal/common_mps.py` own the newly supported var/std dtype and xfail/skip surface; focused `test/test_mps.py` Welford cases own kernel-specific numerical, shape, duplicate-dim, partial-simdgroup, and GroupNorm-relevant regressions; the Inductor MPS selectors own the stale expected-failure removal.

## Scope Notes

- No shared `NormParams` widening is included.
- The only Inductor test metadata change is removing the stale MPS expected-failure marker from `test_tmp_not_defined_issue3`; native `var_mean` support makes the test pass on MPS, and the selector plus its dynamic-shapes variant were verified locally after the removal.
- No generic reduction-dispatch change outside the Welford path is included.
- Benchmark evidence lives in `benchmarks/mps/welford_reduce_bench.py` because it is a standalone MPS reviewer-evidence script using `torch.utils.benchmark`, not an `operator_benchmark` framework case.

## Performance / Memory Stability

Benchmark script: `benchmarks/mps/welford_reduce_bench.py`.

Methodology: same script on the parent build (`1dc5b578e448e411c854983fd1ee983a1d6a80eb`, MPSGraph var/std path) and this head (`cd01523561027511bffff9c99b844a65e693be7b`, native Welford path). Steady rows report mean microseconds over five warm `torch.utils.benchmark.Timer.blocked_autorange` trials and synchronize MPS inside the timed statement. Shape-varying/RSS rows visit 200 distinct shapes to expose graph-cache behavior.

### Steady-State Var

| Shape | Dim | Parent us | This PR us | Speedup |
| --- | ---: | ---: | ---: | ---: |
| `(16777216,)` | `None` | 758.3 | 436.5 | 1.74x |
| `(4096, 4096)` | 1 | 747.7 | 406.4 | 1.84x |
| `(4096, 4096)` | 0 | 787.2 | 421.2 | 1.87x |
| `(1024, 16384)` | 1 | 749.2 | 406.3 | 1.84x |
| `(1000000, 8)` | 1 | 450.5 | 259.2 | 1.74x |
| `(8, 1000000)` | 1 | 426.0 | 327.8 | 1.30x |
| `(8, 1000000)` | 0 | 582.9 | 435.7 | 1.34x |

### Steady-State Std

| Shape | Dim | Parent us | This PR us | Speedup |
| --- | ---: | ---: | ---: | ---: |
| `(16777216,)` | `None` | 757.6 | 426.7 | 1.78x |
| `(4096, 4096)` | 1 | 742.5 | 406.5 | 1.83x |
| `(4096, 4096)` | 0 | 783.0 | 418.6 | 1.87x |
| `(1024, 16384)` | 1 | 747.4 | 403.9 | 1.85x |
| `(1000000, 8)` | 1 | 448.8 | 258.9 | 1.73x |
| `(8, 1000000)` | 1 | 423.0 | 326.5 | 1.30x |
| `(8, 1000000)` | 0 | 599.9 | 434.7 | 1.38x |

### Shape-Varying / RSS

| Op | Parent shape-varying | This PR shape-varying | Parent RSS growth | This PR RSS growth |
| --- | ---: | ---: | ---: | ---: |
| `var` | 878 ms / 200 shapes (4.39 ms/shape) | 6 ms / 200 shapes (0.03 ms/shape) | +2370 MB | +0 MB |
| `std` | 760 ms / 200 shapes (3.80 ms/shape) | 6 ms / 200 shapes (0.03 ms/shape) | +437 MB | +0 MB |

## Test Plan

Current local gate head: `cd01523561027511bffff9c99b844a65e693be7b`.

```bash
/opt/homebrew/bin/python3.11 -m py_compile benchmarks/mps/welford_reduce_bench.py test/test_mps.py torch/testing/_internal/common_methods_invocations.py torch/testing/_internal/common_mps.py
/opt/homebrew/bin/lintrunner -a -m 1dc5b578e448e411c854983fd1ee983a1d6a80eb
# Metal compiler preflight for changed ReduceOps.metal under metal3.1 and metal4.0

rm -f build/caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mps/operations/ReduceOps.mm.o
ninja -C build torch_cpu && cp build/lib/libtorch_cpu.dylib torch/lib/libtorch_cpu.dylib

PYTHONPATH=$PWD PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 /opt/homebrew/bin/python3.11 test/test_mps.py TestMPS.test_std TestReduceOpsMetal -v
PYTHONPATH=$PWD PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 /opt/homebrew/bin/python3.11 test/test_mps.py TestMPS.test_var_simple -v
PYTHONPATH=$PWD PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 /opt/homebrew/bin/python3.11 test/test_mps.py \
  TestConsistencyMPS.test_output_match_var_mps_float32 \
  TestConsistencyMPS.test_output_match_var_mps_complex64 \
  TestConsistencyMPS.test_output_match_std_mps_float32 \
  TestConsistencyMPS.test_output_match_std_mps_complex64 \
  TestConsistencyMPS.test_output_match_var_mean_mps_float32 \
  TestConsistencyMPS.test_output_match_var_mean_mps_complex64 \
  TestConsistencyMPS.test_output_match_std_mean_mps_float32 \
  TestConsistencyMPS.test_output_match_std_mean_mps_complex64 \
  TestConsistencyMPS.test_output_grad_match_var_mps_float32 \
  TestConsistencyMPS.test_output_grad_match_std_mps_float32 \
  TestConsistencyMPS.test_output_grad_match_var_mean_mps_float32 \
  TestConsistencyMPS.test_output_grad_match_std_mean_mps_float32 \
  TestConsistencyMPS.test_output_match_nn_functional_group_norm_mps_float32 \
  TestConsistencyMPS.test_output_grad_match_nn_functional_group_norm_mps_float32 \
  TestCommonMPS.test_numpy_ref_mps_nn_functional_group_norm_mps_float32 \
  -v
PYTHONPATH=$PWD PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 /opt/homebrew/bin/python3.11 -m pytest -q \
  test/inductor/test_torchinductor.py::GPUTests::test_tmp_not_defined_issue3_mps \
  test/inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesGPUTests::test_tmp_not_defined_issue3_dynamic_shapes_mps

PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 PYTHONPATH=/Users/cameronbeeley/pytorch-prs/varstd-rebuild-187787 /opt/homebrew/bin/python3.11 test/run_test.py --verbose --mps --keep-going
```

Full `--mps` local CI result: all PR-relevant var/std, OpInfo/common_mps, `test_mps`, `test_ops`, `test_modules`, and Inductor MPS selectors passed. The runner exited non-zero only because `test/test_metal.py::TestMetalRewritePass::test_conv` fails with `Schema not found for node` for `metal_prepack::conv2d_prepack(...)`; the same selector reproduces on the parent build (`1dc5b578e448e411c854983fd1ee983a1d6a80eb`) with the same error, so this is not a new failure from this PR.

Mutation:

```bash
STRYKER_CXX_BIN=/Users/cameronbeeley/stryker-cxx/bin/stryker-cxx.js \
marmorkrebs --tool stryker-cxx --dir . \
  --changed-files aten/src/ATen/native/mps/operations/ReduceOps.mm:635-690 \
  --max-mutants 12 --threshold-break 0.6 \
  --build-command 'rm -f build/caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mps/operations/ReduceOps.mm.o && ninja -C build torch_cpu && cp build/lib/libtorch_cpu.dylib torch/lib/libtorch_cpu.dylib' \
  --test-command 'PYTHONPATH=$PWD PYTORCH_TESTING_DEVICE_ONLY_FOR=mps PYTORCH_TEST_WITH_SLOW=1 /opt/homebrew/bin/python3.11 .pr-gates/artifacts/welford_complex_empty_check.py'
```

Mutation result: canonical `marmorkrebs` scalar/complex/empty slice scored 0.92 (`12` mutants, `11` killed, `1` equivalent scalar-recursion survivor, threshold `high`). Direct `stryker-cxx` Welford host-line evidence is also preserved: broad lines 368-690 (`56` mutants, `27` killed, remaining survivors dispositioned as equivalent/static-domain/route-perf noise), supplemental dim/shape lines 368-416 (`11` mutants, `6` killed; max-rank and output-stride loop mutants killed), and supplemental complex/empty lines 635-690 (`12` mutants, `11` killed).

## Review History

Review history and per-finding dispositions: https://gist.github.com/2b6a49d1102755c253b0eabb4d2f1694

## BC-breaking?

No. Unsupported `>2^32` MPS var/std indexing domains now fail explicitly; supported dtype/shape behavior is covered against CPU and OpInfo/common_mps.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
Are all of your PRs reuse a lots of the same code from each other? 
Please move the uint32 to int64 migration into a separate PR, but I would propose to measure the perf impact first on the existing ops, as it'll be quite substantial especially for the smaller tensors

**Comment by anagnorisis2peripeteia** [comment]
Thanks, this has been actioned in the rebuilt branch at `cd01523561027`.

The old shared `uint32 -> int64` reduction-indexing migration has been removed from this PR. The rebuilt var/std branch now keeps indexing in the existing uint32 `NormParams` domain and adds explicit unsupported-size errors for `>2^32` input/output/reduction cases instead of widening the shared reduction infrastructure here.

The shared-code split is now:

- #188156: reduce-family base (`value_reduction<Op, Load, TA>`) for sum-family routes.
- #188333: prod on top of that base.
- #187787: Welford-specific var/std kernels only, with OpInfo/common_mps updates and focused MPS/consumer coverage.

I also dropped the earlier widening direction after A/B measurement showed small-tensor regressions, so the perf-risky migration is not part of this PR.

**Comment by anagnorisis2peripeteia** [comment]
This PR is still needed: `var`/`std` still use `LookUpOrCreateCachedGraph` (`varianceOfTensor:`) on current `main`.

It will be rebuilt on current `main` after #188333 (prod) direction is settled, not rebased onto the closed reduce-base #188156. The current commits are stale against Isalia20's faster-reductions rewrite of `ReduceOps.metal`.

Part of #187455.

**Comment by anagnorisis2peripeteia** [comment]
Closing this PR as a dead stack, not as giving up on native var/std.

GitHub is showing four commits:

1. the two #188156 reduce-base commits (closed as superseded by Isalia20 #191101–#191100)
2. the old #188333 prod commit (replaced by the rebuild on current `main`)
3. the Welford commit

The Welford kernels were written against that old `ReduceOps.metal`. They cannot be rebased onto current `main` as a small conflict fix. `var`/`std` still use `LookUpOrCreateCachedGraph` (`varianceOfTensor:`), so the work remains on #187455 and will be a new PR on current `main` after #188333 (prod).
